src/main/java/org/cloudfoundry/promregator/lite/discovery/SimpleTargetResolver.kt

```diff
@@ -0,0 +1,191 @@
+    }
+
+    private suspend fun List<OrgResponse>.resolveSpaces(configTargets: List<Target>, domains: List<DomainResponse>) = coroutineScope {
+        return@coroutineScope this@resolveSpaces.map { org ->
+            async {
+                cfAccessor.retrieveSpaceIdsInOrg(org.api, org.id).map { space ->
+                    SpaceResponse(org, space.entity.name, space.metadata.id, domains.filter { it.orgId == org.id })
+                }
+            }
+        }.awaitAll().flatten().filter { space ->
+            configTargets.any { target ->
+                (target.spaceName == null && target.spaceRegex == null) ||
+                        (target.spaceName == null && target.spaceRegex?.toRegex(IGNORE_CASE)?.matches(space.spaceName) == true) ||
+                        (target.spaceName == space.spaceName && target.spaceRegex == null)
+            }
+        }
+    }
+
+    private suspend fun List<SpaceResponse>.resolveSummaries() = coroutineScope {
+        return@coroutineScope this@resolveSummaries.map { space ->
+            async {
+                cfAccessor.retrieveSpaceSummary(space.org.api, space.spaceId)?.applications?.map { app ->
+                    val appName = app.name.lowercase(LOCALE_OF_LOWER_CASE_CONVERSION_FOR_IDENTIFIER_COMPARISON)
+                    SpaceSummaryAppResponse(space, appName, AppSummary(app.instances, app.urls, app.routes))
+                } ?: listOf()
+            }
+        }.awaitAll().flatten()
+    }
+
+    private suspend fun List<SpaceResponse>.resolveAppsV3(spaceSummaries: List<SpaceSummaryAppResponse>) = coroutineScope {
+        return@coroutineScope this@resolveAppsV3.map { space ->
+            async {
+                val appsv3 = cfAccessor.retrieveAllApplicationsInSpaceV3(space.org.api, space.org.id, space.spaceId)
+                appsv3.map { appRes ->
+                    val scrapable = isApplicationInScrapableState(appRes.state.toString())
+                    val annotationScrape = appRes.metadata.annotations.getOrDefault(PROMETHEUS_IO_SCRAPE, "false") == "true"
+                    val annotationMetricsPath = appRes.metadata.annotations[PROMETHEUS_IO_PATH]
+
+                    /*
+                     * Due to https://github.com/cloudfoundry/cloud_controller_ng/issues/1523, we
+                     * cannot rely on sas.getId() (i.e. it may contain wrong information) So we use the name
+                     */
+                    val appNameLower = appRes.name.lowercase(LOCALE_OF_LOWER_CASE_CONVERSION_FOR_IDENTIFIER_COMPARISON)
+                    AppResponse(appRes.id, appRes.name, space, scrapable, annotationMetricsPath,
+                            spaceAppSummary = spaceSummaries.firstOrNull { it.appName == appNameLower }?.appSummary)
+                }
+
+            }
+        }.awaitAll().flatten()
+    }
+
+    private suspend fun List<SpaceResponse>.resolveAppsV2() = coroutineScope {
+        return@coroutineScope this@resolveAppsV2.map { space ->
```

Comment:
These `return@coroutineScope` syntax aren't required. I included them to help explain what is being returned for those new to Kotlin. We can remove them